### DISCUSSION TITLE
Most Efficient Way to Retrieve a Large Quantity of Quotes

### DISCUSSION THREAD

**Discussion by def-mycroft** [erdewit/ib_insync#467] [Q&A]
Hello, I want to retrieve market quotes and recent trades for a large number of contracts; but after using `ib_insync` for a while I've not yet figured out how to accomplish this efficiently. 

I want to be able to pull down current prices and use them as inputs to my model; ideally I'd be able to efficiently get current prices for ~10-15K individual options contracts (on the order of 35 strikes for each of 3 expiration cycles for each of 100 underlying assets). 

I don't require the data at a high frequency, I just need to refresh data at maximum a few times per day; therefore I'd be satisfied if it 30-60 minutes to complete. 

Using `ib.reqHistoricalData` in a loop for each contract is too slow (at least in the way I'm implementing it); in what way might I be able to implement differently that would scale to ~10-15K contracts? 

I'm not necessarily looking for a code recipe (but that would be much appreciated); just getting a direction would be helpful. I'm not familiar with `async`, therefore I'm not sure if that would be helpful in this case or not. 

Many thanks for any help.

**Comment by mattsta** [comment]
There's not really a good answer for bulk requesting historical data from ibkr because they aggressively throttle and rate limit all their data endpoints.

See https://interactivebrokers.github.io/tws-api/historical_limitations.html and https://interactivebrokers.github.io/tws-api/contract_details.html (especially the note about "a similar subsequent one will be kept on hold for one minute" — they rate limit some requests to one fetch *per minute*!).

The only reasonable bulk APIs I've seen for options are https://polygon.io/pricing (no rate limits, but can get expensive) and https://documentation.tradier.com/brokerage-api/markets/get-quotes (free with account, rate limited to 120 requests per minute which is usually enough).
